src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -37,7 +37,11 @@ private class CallbackRunnable[T](val executor: ExecutionContext, val onComplete
     value = v
     // Note that we cannot prepare the ExecutionContext at this point, since we might
     // already be running on a different thread!
-    try executor.execute(this) catch { case NonFatal(t) => executor reportFailure t }
+    try executor.execute(this) catch { case NonFatal(e) =>
+      executor reportFailure e
+      value = Failure(e)
+      run()
```

Comment:
I am not convinced this is safe. If `executor.execute(this)` is executed synchronously then this may cause some callbacks to be executed with the first value, and then others with the Failure(e). This would violate Future semantics.

Also, it would execute the run() in the calling thread.